numba/jitclass/boxing.py

```diff
@@ -0,0 +1,196 @@
+    assert not argspec.varargs, 'varargs not supported'
+    assert not argspec.keywords, 'keywords not supported'
+    assert not argspec.defaults, 'defaults not supported'
+
+    args = ', '.join(argspec.args[1:])  # skipped self arg
+    source = _method_code_template.format(method=name, args=args)
+    glbls = {}
+    exec_(source, glbls)
+    method = njit(glbls['method'])
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        return method(*args, **kwargs)
+
+    return wrapper
+
+
+_cache_specialized_box = {}
+
+
+def _specialize_box(typ):
+    """
+    Create a subclass of Box that is specialized to the jitclass.
+
+    This function caches the result to avoid code bloat.
+    """
+    # Check cache
+    if typ in _cache_specialized_box:
+        return _cache_specialized_box[typ]
+    dct = {'__slots__': ()}
+    # Inject attributes as class properties
+    for field in typ.struct:
+        if not field.startswith('_'):
+            getter = _generate_getter(field)
+            setter = _generate_setter(field)
+            dct[field] = property(getter, setter)
+    # Inject methods as class members
+    for name, func in typ.methods.items():
+        if not name.startswith('_'):
+            getter = _generate_method(name, func)
+            dct[name] = getter
+    # Create subclass
+    subcls = type(typ.classname, (Box,), dct)
+    # Store to cache
+    _cache_specialized_box[typ] = subcls
+
+    return subcls
+
+
+class Box(object):
+    """
+    A box for numba created jit-class instance
+    """
+    __slots__ = '_meminfo', '_meminfoptr', '_dataptr', '_numba_type_'
+
+    def __init__(self, meminfoptr, dataptr, typ):
+        # MemInfo is used to acquire a reference to `meminfoptr`.
+        # When the MemInfo is destroyed, the reference is released.
+        self._meminfo = MemInfo(meminfoptr)
+        self._meminfoptr = meminfoptr
+        self._dataptr = dataptr
+        self._numba_type_ = typ
+
+
+###############################################################################
+# Implement box/unbox for call wrapper
+
+@box(types.ClassInstanceType)
+def _box_class_instance(typ, val, c):
+    meminfo, dataptr = cgutils.unpack_tuple(c.builder, val)
+
+    lluintp = c.context.get_data_type(types.uintp)
+
+    addr_meminfo = c.pyapi.from_native_value(types.uintp,
+                                             c.builder.ptrtoint(meminfo,
+                                                                lluintp))
+    addr_dataptr = c.pyapi.from_native_value(types.uintp,
+                                             c.builder.ptrtoint(dataptr,
+                                                                lluintp))
+
+    # XXX: relies on runtime address
+    int_addr_typ = c.context.get_constant(types.uintp, id(typ))
+
+    box_subclassed = _specialize_box(typ)
+    # Note: the ``box_subclassed`` is kept alive by the cache
+    int_addr_boxcls = c.context.get_constant(types.uintp, id(box_subclassed))
```

Comment:
... but of course, there's still a raw pointer constant here. I can't think of any easy solution, apart from merging Box into the MRO of the class returned by `@jitclass`.